beacon_node/beacon_chain/src/block_status_table.rs

```diff
@@ -0,0 +1,784 @@
+use dashmap::DashMap;
+use dashmap::mapref::entry::Entry;
+use std::sync::atomic::{AtomicU8, AtomicUsize, Ordering};
+use tokio::sync::broadcast;
+use types::Hash256;
+
+/// Represents the status of a block during the data availability import process.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[repr(u8)]
+pub enum BlockStatus {
+    /// Block is pending processing
+    Pending = 0,
+    /// Block is currently being executed
+    Executing = 1,
+    /// Block has been determined to be invalid
+    Invalid = 2,
+    /// Block is currently being imported
+    Importing = 3,
+    /// Block has been successfully imported
+    Imported = 4,
+}
+
+impl BlockStatus {
+    /// Converts a u8 value to a Status enum.
+    /// Returns None if the value is invalid.
+    pub fn from_u8(value: u8) -> Option<Self> {
+        match value {
+            0 => Some(BlockStatus::Pending),
+            1 => Some(BlockStatus::Executing),
+            2 => Some(BlockStatus::Invalid),
+            3 => Some(BlockStatus::Importing),
+            4 => Some(BlockStatus::Imported),
+            _ => None,
+        }
+    }
+
+    /// Converts the Status to its u8 representation.
+    pub fn to_u8(self) -> u8 {
+        self as u8
+    }
+
+    /// Returns true if the status represents an active processing state
+    pub fn is_active(&self) -> bool {
+        matches!(self, BlockStatus::Executing | BlockStatus::Importing)
+    }
+
+    /// Returns true if the status represents a terminal state
+    pub fn is_terminal(&self) -> bool {
+        matches!(self, BlockStatus::Invalid | BlockStatus::Imported)
+    }
+}
+
+/// A thread-safe status tracking table for Ethereum blocks during data availability import.
+///
+/// This table uses DashMap for lock-free concurrent access and AtomicU8 values for
+/// efficient atomic operations on status transitions to prevent race conditions during
+/// block processing.
+#[derive(Debug)]
+pub struct BlockStatusTable {
+    /// The underlying concurrent hash map storing block status as AtomicU8
+    table: DashMap<Hash256, AtomicU8>,
+    /// Broadcast channels for notifying subscribers of status changes
+    notifiers: DashMap<Hash256, broadcast::Sender<BlockStatus>>,
+    /// Counter for tracking the number of entries (for metrics/debugging)
+    entry_count: AtomicUsize,
```

Comment:
Should be the initial design be the simplest approach possible?

```rust
pub enum BlockStatus {
    Pending,
    Executing(broadcast::Sender<BlockStatus>),
    Invalid,
    Importing(broadcast::Sender<BlockStatus>),
    Imported.
}

pub struct BlockStatusTable {
    table: DashMap<Hash256, BlockStatus>,
}
```

We can optimize latter with more dashmaps if we happen to observe and measure performance degradation due to lock contention